[flang] Add design document for debug info generation.

abidh opened the pull request:
This document discusses some options where the debug metadata can be generated.
 It also goes through various language constructs and explains how the debug metadata
 will look like for that construct and how we can extract that information.

The real point of discussion is how and where to extract the information about various
language features to generate the debug metadata. The structure of the metadata itself
is mostly settled as that is dictated by the DWARF and structure of LLVM IR metadata.
The classic flang and gfortran generate quite similar DWARF for the various language constructs.

This document is based on what Kiran posted in https://reviews.llvm.org/D138534.

abidh committed:
[flang] Add design document for debug info generation.

This document discusses some options where the debug metadata can be generated.
 It also goes through various language constructs and explains how the debug metadata
 will look like for that construct and how we can extract that information.

abidh committed:
[flang][docs] Add reference to DebugGeneration in toctree.

tblah reviewed (commented):
Thank you very much for writing this comprehensive design document.

I have a few questions but on the whole I like the approach here.

abidh committed:
Apply suggestions from code review

Co-authored-by: Tom Eccles <t@freedommail.info>

abidh commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,452 @@
+
+  In FIR, `DeclareOp` has source information about the variable. The
+  `DeclareOp` will be processed to create `DILocalVariableAttr`. This attr is
+  attached to the memref op of the `DeclareOp` using a `FusedLoc` approach.
```
I agree that it make more sense on `DeclareOp` conceptually but it does makes the implementation more tricky. It will at least require us to maintain a mapping from the memref op in FIR to converted one in LLVM dialect.

jeanPerier commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,452 @@
+#### Assumed Size
+
+This is treated as raw array. Debug information will not provide any bounds
+information.
```
> will not provide any bounds information.

Is this true for any bounds or only the upper bound in the last dimension?

abidh replied:
Upper bound in last dimension.

jeanPerier commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,452 @@
+
+### Namelists
+
+FIR does not seem to have a way to extract information about namelists.
```
Namelists are handled by creating [runtime data structures](https://github.com/llvm/llvm-project/blob/cb9cf331fa69bb8319886cdca043a742b4d19ce5/flang/runtime/namelist.h#L30) that describes them. These data structures are globals when possible, and locals otherwise (without fir.declare currently). They can be identified via [their mangling](https://github.com/llvm/llvm-project/blob/main/flang/docs/BijectiveInternalNameUniquing.md#namelist-groups) using [InternalNames.h](https://github.com/llvm/llvm-project/blob/main/flang/include/flang/Optimizer/Support/InternalNames.h).

It may be non trivial to generate the debug info from the global/local. Maybe we could abstract that more in FIR/Lowering.

abidh replied:
Thanks for the information. Although when I tried a simple namelist example, I did not see a local or global with this name mangling scheme in fir. Ideally we will need something that represent namelist and then refer to `DeclareOp` of all the variables that are part of the given namelist.

jeanPerier replied:
These data structures are created lazily when encountering the namelist in an IO statement, so if your test had no IO statements using it, you would indeed not have seen it. I understand this is not convenient from a debug info point of view, again more abstraction might be needed in lowering here to generate the debug info.

jeanPerier commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,452 @@
+
+They will be treated like normal variables. Although we may require to handle
+the case where the `DeclareOp` of one variable points to the `DeclareOp` of
+another variable (e.g. a => b).
```
Currently inlining is not done in FIR, but that would also create DeclareOp chains.

jeanPerier commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,452 @@
+available in RecordType or TypeInfoOp. This probably will need to be
+added.
+
+### CommonBlocks
```
Do you know if something special is needed to deal with equivalence outside of Common Blocks, or can the output of the fir.coordinate_of be used for the `DILocalVariableAttr`?

e.g:

```
subroutine test_locals()
  real :: x(100), y
  equivalence (x(10), y)
end subroutine
```

abidh replied:
Thanks for pointing out this case. The `DbgDeclareOp` takes a 3rd expression argument which we can use to handle cases like this by using something like `DIExpression(DW_OP_plus_uconst, 36)` for the example case.

jeanPerier reviewed (commented):
Thanks for writing this down, this very useful!
I am in line with the approach. Few questions/notes below.

abidh commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,452 @@
+
+3. The location where a component of the derived type is declared is not
+available in RecordType or TypeInfoOp. This probably will need to be
+added.
```
I think it can work. I looped over all components in createTypeInfoOp and added the locations as FusedLoc. Later in the AddDebugInfo pass, I can see the following.

`fir.type_info @_QMhelperTt_pair noinit nodestroy nofinal : !fir.type<_QMhelperTt_pair{i:i32,i1:i32,x:f32}> loc(fused<loc("derived.f90":6:13)>[fused<loc("derived.f90":5:16)>[fused<loc("derived.f90":4:16)>["derived.f90":3:11]]])`

abidh commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,452 @@
+In FIR, the `DeclareOp` points to a `ShapeOp` and we can walk the chain
+to get the value that represents the array bound in any dimension. We will
+create a `DILocalVariableAttr` that will point to that location. This
+variable will be used in the `DISubrangeAttr`. Note that this
```
The first argument to `DebugDeclareOp` needs to be an llvm pointer type so an SSA value may not work. I looked at the IR from the classic flang and it does seem to put such expressions to memory.

jeanPerier replied:
Could llvm.dbg.value be used instead of llvm.dbg.declare here?
If not, I guess the debug info pass will have to insert allocas for bounds that are non constant SSA values (even if lowering did lowered them in memory, we could not guarantee they would not be promoted to values by some MLIR optimization).

abidh replied:
The llvm.dbg.value could work. As you mentioned, there is option of adding allocas during debug info generation. The issue may be to keep the value alive throughout the function.

jiel-nv commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,452 @@
+
+### Strings
+
+Fixed sized strings will be treated like fixed sizes arrays and
```
Starting from LLVM 12, the Fortran string types (fixed length, assumed length, deferred length) LLVM debug metadata are represented by `DIStringType` (with `name`, `size`, `stringLength`, `stringLengthExpression` fields).

abidh replied:
Thanks for pointing that out. That approach would require adding an artificial variable for string length but should be relatively straight forward to implement. I will update the document and add the example IR. We will also need to add `DIStringTypeAttr` in the MLIR.

abidh committed:
Handle code review comments.

1. Mention the upper bound of the last dimension in
   the 'Assumed Size' section.
2. Redo 'Strings' section using 'DIStringType'.
3. Mention 'DIStringType' in 'Missing metadata in MLIR'

jeanPerier reviewed (approved):
Thanks for the document and the answers to the comments. Looks great!

tblah reviewed (approved):
LGTM, thanks for all of your work on this!

kiranchandramohan commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,451 @@
+  - Break at functions.
+  - print type (ptype) of function names.
+  - print values and types (ptype) of various type of variables
+- Manually run `GDB`'s gdb.fortran testsuite with llvm-flang.
```
Can we add the `gdb.fortran` tests to the llvm-testsuite just like the gfortran tests are added?

abidh replied:
The `GDB` tests require its tcl based framework to issue commands and look for results. It may not be feasible to add them to llvm-testsuite. Adding a bot that builds flang and then run `gdb.fortran` tests with it may be a possibility.

kiranchandramohan commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,451 @@
+- Calling Convention: `DW_CC_normal` by default and `DW_CC_program` if it is
+the main program.
+
+`DISubroutineTypeAttr` currently has a fixed type. This will be changed to
+match the signature of the actual function/subroutine.
```
How do you plan to get the actual signature since this is going to be different from what is there in FIR?

abidh replied:
What is meant by actual signature and how this will be different to the information that is in the FIR. My plan was to create the `DISubroutineTypeAttr` based on the information about arguments and results in `FuncOp`. A few simple cases that I tried, the results of the `ptype` command of the `GDB` were similar for executable generated with `flang-new` and `gfortran`.

kiranchandramohan replied:
As discussed in https://reviews.llvm.org/D138534, for the following program, what would be the `ptype` for the function foo?
```
 program mn
  integer :: arr(10), cnt
  call foo(arr, cnt)
contains
  subroutine foo( a, n)
    integer  :: a(:), n
  end subroutine foo
end program
```
gfortran (Actual signature)
```
ptype foo
type = void (integer(kind=4) (:), integer(kind=4))
```
Classic Flang (LLVM IR signature)
```
(gdb) ptype foo
type = void (integer*8, integer, integer (187651372366496), uinteger*8)
```

abidh replied:
I expect it to be similar to gfortran.

kiranchandramohan commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,451 @@
+The assumed shape array will use the similar representation as fixed size
+array but there will be 2 differences.
+
+1. There will be a `datalocation` field which will be an expression. This will
+enable debugger to get the data pointer from array descriptor.
+
+2. The field in `DISubrangeAttr` for array bounds will be expression which will
+allow the debugger to get the bounds from descriptor.
```
Will all these be generated during FIRToLLVM conversion? How will you get the right offsets from the descriptor?

abidh replied:
These will be generated during normal debug attribute generation in AddDebugInfoPass. I am not clear about the 2nd part of the question. If you mean how we will get the offsets of this information in descriptor, I think we know the layout of the descriptor in the memory and can tell where various things are located. If you mean how debugger will get this information, we will encode them in `DWARF`'s expressions which can be decoded by the debuggers.

Please note that most of the examples that I included in the document are known to work. Here are the steps I followed to verify the example debug metadata.

1. Use flang-new to create LLVM IR file from the sample fortran file.
2. Manually edit the .ll file to add the example debug metadata
3. Build the edited file into an executable
3. Run the executable under `GDB` and check that debugger is able to extract the right information.

kiranchandramohan replied:
>  If you mean how we will get the offsets of this information in descriptor
It is this part. The offsets are only available when the descriptor is converted to a structure in the FIRToLLVM conversion pass. So how can you get this info in the AddDebugInfo pass?

abidh replied:
I thought the layout was known. For example, the FIRToLLVM seems to rely on the hardcoded values like `kDimsPosInBox`. We could have similar thing in AddDebugInfo pass but with offsets. 

I also inspected the layout of the descriptor at runtime in the debugger and then looked at the CFI_cdesc_t and it matched.

kiranchandramohan commented on flang/docs/DebugGeneration.md:
````diff
@@ -0,0 +1,451 @@
+!8 = !DIBasicType(tag: DW_TAG_base_type, name: "integer" ...)
+```
+
+In assumed shape case, the rank can be determined from the FIR's `SequenceType`.
+This allows us to generate a `DISubrangeAttr` in each dimension.
````
Is this also available from the descriptor?

abidh replied:
I mean to say that `DISubrangeAttr` can encode where the information about the array bounds is located in the descriptor for each dimension.

kiranchandramohan reviewed (approved):
Nice work. Thanks for the writing this down and the initial work done for this. I have a few questions. But no need to wait for further replies from my side.

abidh committed:
Update flang/docs/DebugGeneration.md

Review Suggestion.

Co-authored-by: Kiran Chandramohan <kiranchandramohan@gmail.com>

abidh commented on flang/docs/DebugGeneration.md:
```diff
@@ -0,0 +1,451 @@
+
+The AddDebugFoundationPass will be renamed to AddDebugInfo Pass. The
+information mentioned in the line info section above will be passed to it from
+the driver. This pass will run quite late in the pipeline but before
```
Can you kindly give an example of thing which will be present in HLFIRDeclare and not in FIRDeclare. I don't mind using it but then we will have to run the AddDebugInfo pass early when hlfir is still there.

abidh committed:
Remove questions that have been answered during code review.
